src/Pirouette/SMT/Monadic.hs

```diff
@@ -102,38 +103,50 @@ checkSat = do
 solverPush :: (MonadIO m) => SolverT m ()
 solverPush = do
   solver <- ask
-  liftIO $ SimpleSMT.push solver
+  liftIO $ PureSMT.push solver
 
 -- | Pops the current checkpoint, check 'solverPush' for an example.
 solverPop :: (MonadIO m) => SolverT m ()
 solverPop = do
   solver <- ask
-  liftIO $ SimpleSMT.pop solver
+  liftIO $ PureSMT.pop solver
 
 -- | Declare a single datatype in the current solver session.
 declareDatatype :: (LanguageSMT lang, MonadIO m) => Name -> TypeDef lang -> ExceptT String (SolverT m) [Name]
 declareDatatype typeName (Datatype _ typeVariables _ cstrs) = do
   solver <- ask
   (constr', _) <- runWriterT $ mapM (constructorFromPIR typeVariables) cstrs
   liftIO $ do
-    SimpleSMT.declareDatatype
+    PureSMT.declareDatatype
       solver
       (toSmtName typeName)
       (map (toSmtName . fst) typeVariables)
       constr'
   return $ typeName : map fst cstrs
 
 -- | Declare a set of datatypes in the current solver session, in the order specified by
--- the dependency order passed as the second argument. You can generally get its value
--- from a 'PirouetteDepOrder' monad.
+-- the list of definitions. For info on sorting definitions, check the 'PirouetteDepOrder' class.
 declareDatatypes ::
-  (LanguageSMT lang, MonadIO m) => M.Map Name (Definition lang) -> [R.Arg Name Name] -> ExceptT String (SolverT m) [Name]
-declareDatatypes decls orderedNames = do
-  let tyNames = mapMaybe (R.argElim Just (const Nothing)) orderedNames
-  concatForM tyNames $ \tyname ->
-    case M.lookup tyname decls of
-      Just (DTypeDef tdef) -> declareDatatype tyname tdef
-      _ -> return []
+  (LanguageSMT lang, MonadIO m) => [(Name, TypeDef lang)] -> ExceptT String (SolverT m) [Name]
+declareDatatypes = fmap concat . mapM (uncurry declareDatatype)
+
+-- |If a function can be declared as an uninterpreted function, returns
+-- the type of its arguments and the type of its result.
+supportedUninterpretedFunction ::
```

Comment:
previously, we were using `declareAsManyUninterpretedFunctinosAsPossible` inside `initSolver`, and `initSolver` would return the list of names that managed to be declared. Those that we don't "support", just get silently ignored, which was fine but it adds a dependency between `initSolver` and everything else. 

I split `declareAsManyUninterpretedFunctionsAsPossible` into two: first give me the list of functions that I know I can declared (because we've managed to translate their types already), then declare then in `initSolver`. Now, I don't need the result of `initSolver` to have the set of `knownNames`.

I'll remove the old `declareAsManyUninterpretedFunctionsAsPossible` and will update the comment.

Reply:
(I'll keep this unresolved so the comment is always visible, breaking the data dependencies is an important point of this PR)